File: src/main/java/org/hildan/minecraft/mining/optimizer/ore/MathHelper.java
package org.hildan.minecraft.mining.optimizer.ore;

@SuppressWarnings({"PointlessBitwiseExpression", "MagicNumber"})
class MathHelper {

    private static final int N_STEPS = 65536;

    private static final float[] SIN_CACHE = new float[N_STEPS];

    static {
        for (int i = 0; i < SIN_CACHE.length; ++i) {
            SIN_CACHE[i] = (float) Math.sin(i * Math.PI * 2 / N_STEPS);
        }
    }

    public static final float sin(float f) {
        return SIN_CACHE[(int) (f * 10430.378F) & 0xffff];
    }

    public static final float cos(float f) {
        return SIN_CACHE[(int) (f * 10430.378F + 16384.0F) & 0xffff];
    }

    public static int floor(double d0) {
        int i = (int) d0;
        return d0 < i ? i - 1 : i;
    }
}

File: src/main/kotlin/org/hildan/minecraft/mining/optimizer/ore/MathHelper.kt
package org.hildan.minecraft.mining.optimizer.ore

internal object MathHelper {

    private const val N_STEPS = 65536

    private val SIN_CACHE = (0 until N_STEPS).map { Math.sin(it * Math.PI * 2 / N_STEPS) }

    fun sin(f: Double) = SIN_CACHE[(f * 10430.378).toInt() and 0xffff]

    fun cos(f: Double) = SIN_CACHE[(f * 10430.378 + 16384.0).toInt() and 0xffff]

    fun floor(d: Double): Int {
        val i = d.toInt()
        return if (d < i) i - 1 else i
    }
}
